p4_symbolic/ir/ir.proto

```diff
@@ -0,0 +1,94 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// The main protobuf file for definition of our intermediate representation.
+// This file defines the overall structure of a P4 program in our IR.
+// The structure is tailored to be easy to consume rather than produce.
+// It deviates from the underlying bmv2 JSON format, which is mimiced closely
+// by p4_symbolic/bmv2/bmv2.proto.
+// The bmv2 proto is used to initially parse the JSON format, which is then
+// processed by our tool and transformed into the IR.
+// The IR does not cover all the features of P4 programs.
+// Specifically, we intentionally ignore parsers, de-parsers, stacks and unions,
+// checksum calculations, and extern definitions.
+// Additionally, some types of expressions and statements are unsupported, these
+// are described in values.proto and statements.proto respectively.
+
+syntax = "proto3";
+
+package p4_symbolic.ir;
+
+import "p4_pdpi/ir.proto";
+import "p4_symbolic/bmv2/bmv2.proto";
+// TODO(babman): should we combine all the .proto files? discuss.
+import "p4_symbolic/ir/statements.proto";
+
+// Overall structure of a p4 program.
+message P4Program {
+  // Header definitions.
+  map<string, HeaderType> headers = 1;
+  // Action definitions (including action bodies), keyed by action name.
+  map<string, Action> actions = 2;
```

Comment:
I thought we wanted to rely mostly on the IR from pdpi, and only add what is missing. So for actions, I would expect something like

map<string, pdpi.ir.IrActionDefinition> actions;
map<string, ActionImplementation> action_implementations;

The former gives you the definition, and the latter has only what the former does not have, which is basically the body of the action.

The proto as it is right now duplicates everything between actions and actions_pdpi, if I see correctly.

Reply:
Yes, you are right about the duplicate. I was thinking of doing something similar to what you describe. There are two options that we have, maybe they are very similar and it does not really matter. 

1) doing exactly what you said, where the pdpi definition is added at the top level as a collection over all actions, and then our own specialized implementations are added as a collection as well.

2) Creating a protobuf type that combines pdpi and implementation of a single action, then having a collection of that type (instead of 2 separate collections). I was going to implement this second option next.

Maybe these are very similar. My thought process is that (2) is easier to consume since all information about a single action is in one object (instead of two objects in two sibling maps) but (1) is easier to produce.

Reply:
Yeah sorry, 2 probably makes more sense.  Basically you are saying there will be something like this:

map<string, Action> actions;

message Action {
pdpi.ir.IrActionDefinition definition;
ActionImplementation implementation;
}

Reply:
exactly, will do this and push the update next.

Reply:
done